### PR TITLE
test: Improve tests for legacy omnichannel analytics 

### PR DESCRIPTION
**Pull request by murtaza98**
## Issue(s)

https://rocketchat.atlassian.net/browse/OC-1272

### CODE REVIEW

File: apps/meteor/tests/data/livechat/utils.ts [hunk 1]
```diff
@@ -4,3 +4,5 @@ export type DummyResponse<T, E = 'wrapped'> =
 export const sleep = (ms: number) => {
 	return new Promise((resolve) => setTimeout(resolve, ms));
 }
+
+export const random = (min: number, max: number) => Math.floor(Math.random() * (max - min + 1) + min);
```
Comment (KevLehman) [inline apps/meteor/tests/data/livechat/utils.ts:8]: Can't u use the random package?

File: apps/meteor/app/livechat/server/methods/getAgentOverviewData.ts [hunk 2]
```diff
@@ -18,9 +18,7 @@ declare module '@rocket.chat/ui-contexts' {
 
 Meteor.methods<ServerMethods>({
 	async 'livechat:getAgentOverviewData'(options) {
-		methodDeprecationLogger.warn(
-			'The method "livechat:getAgentOverviewData" is deprecated and will be removed after version v7.0.0. Use "livechat/analytics/agent-overview" instead.',
-		);
+		methodDeprecationLogger.method('livechat:getAgentOverviewData', '7.0.0', ' Use "livechat/analytics/agent-overview" instead.');
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/methods/getAgentOverviewData.ts:21]: Move this to another PR

File: apps/meteor/app/livechat/server/methods/getAnalyticsOverviewData.ts [hunk 3]
```diff
@@ -18,6 +19,7 @@ declare module '@rocket.chat/ui-contexts' {
 
 Meteor.methods<ServerMethods>({
 	async 'livechat:getAnalyticsOverviewData'(options) {
+		methodDeprecationLogger.method('livechat:getAnalyticsOverviewData', '7.0.0', ' Use "livechat/analytics/overview" instead.');
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/methods/getAnalyticsOverviewData.ts:22]: Move this to another PR :)

(Both these should be on a changeset)

File: apps/meteor/tests/data/livechat/department.ts [hunk 4]
```diff
@@ -59,29 +58,19 @@ new Promise((resolve, reject) => {
 
 export const createDepartmentWithAnOnlineAgent = async (): Promise<{department: ILivechatDepartment, agent: {
 	credentials: IUserCredentialsHeader;
-	user: IUser;
+	user: IUser & { username: string };
```
Comment (KevLehman) [inline apps/meteor/tests/data/livechat/department.ts:61]: What about adding a type like this (in case we don't have it)
```
type WithRequiredProperty<Type, Key extends keyof Type> = Type & {
  [Property in Key]-?: Type[Property];
};

WithRequiredProperty<IUser, 'username'>
```

File: apps/meteor/tests/data/livechat/department.ts [hunk 5]
```diff
@@ -59,29 +58,19 @@ new Promise((resolve, reject) => {
...
-	const agent = body.user;
-	const createdUserCredentials = await login(agent.username, password);
-	await createAgent(agent.username);
-	await makeAgentAvailable(createdUserCredentials);
+    const { user, credentials } = await createAnOnlineAgent();
 
 	const department = await createDepartmentWithMethod() as ILivechatDepartment;
 
-	await addOrRemoveAgentFromDepartment(department._id, {agentId: agent._id, username: (agent.username as string)}, true);
+	await addOrRemoveAgentFromDepartment(department._id, {agentId: user._id, username: (user.username as string)}, true);
```
Comment (KevLehman) [inline apps/meteor/tests/data/livechat/department.ts:67]: These `xOrY` functions drive me crazy 🙈 we should have one for each case, so it's obvious for you what are you doing.

We can improve that on other PR

Comment (KevLehman) [inline apps/meteor/tests/data/livechat/department.ts:67]:
```suggestion
	await addOrRemoveAgentFromDepartment(department._id, {agentId: user._id, username: user.username }, true);
```

If you're adding `username` as non optional, you shouldn't need this

File: apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts [hunk 6]
```diff
@@ -1,19 +1,143 @@
...
+		await addOrRemoveAgentFromDepartment(department._id, { agentId: agent2.user._id, username: agent2.user.username }, true);
+		agents.push(agent1);
+		agents.push(agent2);
+
+		// start a few chats
+		const { room: r1, visitor: v1 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent1.credentials });
+		const { room: r2, visitor: v2 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent1.credentials });
+		const { room: r3, visitor: v3 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent2.credentials });
+		const { room: r4, visitor: v4 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent2.credentials });
+		const { room: r5, visitor: v5 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent2.credentials });
```
Comment (KevLehman) [inline apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts:109]: `promise.all`

File: apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts [hunk 7]
```diff
@@ -1,19 +1,143 @@
...
+		const room5ChatStart = moment();
+		const { room: r6, visitor: v6 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent2.credentials });
+		const room6ChatStart = moment();
+		const { room: r7, visitor: v7 } = await startANewLivechatRoomAndTakeIt({ departmentId: department._id, agent: agent2.credentials });
+
+		chatInfo.push({ room: r1, visitor: v1, agent: agent1 });
```
Comment (KevLehman) [inline apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts:115]: Instead of pushing, why not define `chatInfo` here with all the elements there already?

File: apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts [hunk 8]
```diff
@@ -101,6 +293,39 @@ describe('LIVECHAT - dashboards', function () {
...
+			const result = await request
+				.get(api('livechat/analytics/dashboards/chats-totalizers'))
+				.query({ start, end, departmentId: department._id })
+				.set(credentials)
+				.expect('Content-Type', 'application/json')
+				.expect(200);
+
+			const expected = [
+				{ title: 'Total_abandoned_chats', value: 1 },
+				{ title: 'Avg_of_abandoned_chats', value: '14%' },
```
Comment (KevLehman) [inline apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts:309]: If another test before this one adds another room, won't this value change?

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/app/livechat/server/methods/getAgentOverviewData.ts (hunk 2 above)
Comment (murtaza98) [reply apps/meteor/app/livechat/server/methods/getAgentOverviewData.ts:21]: Got it :+1: Moved it here https://github.com/RocketChat/Rocket.Chat/pull/30421

File: apps/meteor/app/livechat/server/methods/getAnalyticsOverviewData.ts (hunk 3 above)
Comment (murtaza98) [reply apps/meteor/app/livechat/server/methods/getAnalyticsOverviewData.ts:22]: Moved it here -> https://github.com/RocketChat/Rocket.Chat/pull/30421

File: apps/meteor/tests/data/livechat/utils.ts (hunk 1 above)
Comment (murtaza98) [reply apps/meteor/tests/data/livechat/utils.ts:8]: Good idea!

File: apps/meteor/tests/data/livechat/department.ts (hunk 5 above)
Comment (murtaza98) [reply apps/meteor/tests/data/livechat/department.ts:67]: You mean breaking down this method to something like:
`addAgentToDepartment` and `removeAgentFromDepartment`

File: apps/meteor/tests/data/livechat/department.ts (hunk 4 above)
Comment (murtaza98) [reply apps/meteor/tests/data/livechat/department.ts:61]: Yups nice! Much better :ok_hand:

File: apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts (hunk 6 above)
Comment (murtaza98) [reply apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts:109]: Oops!

File: apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts (hunk 8 above)
Comment (murtaza98) [reply apps/meteor/tests/end-to-end/api/livechat/04-dashboards.ts:309]: We're filtering by departments here, which department was only created for this test-case. Hence another tests won't have any impact of this

**Review by KevLehman — APPROVED** [review]
